### ISSUE TITLE
Size Difference between QDirStat Reports and the "du" and "df" Commands

### ISSUE THREAD

**Issue by shundhammer** [shundhammer/qdirstat#134]
# Problem

QDirStat sometimes reports a different total size than shell commands like `du` ("disk usage"), even when running QDirStat with full root permissions:

```
% sudo du -hsx /
9,4G	/

% df -h /
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdc2        30G  9,5G   19G  34% /

% sudo qdirstat /
```
![QDirStat-root-fs](https://user-images.githubusercontent.com/11538225/84180684-32a26880-aa88-11ea-992f-a7056f41cb63.png)

It's even different from its own "Filesystems" window:

![QDirStat-fs](https://user-images.githubusercontent.com/11538225/84180807-6aa9ab80-aa88-11ea-981e-140e177cddb8.png)

**Comment by shundhammer** [comment]
This is the result of different sizes being used: A file's _byte size_ vs. the _allocated blocks_.

Filesystems use blocks in multiples of 512 bytes internally. So even a file with only 1 byte will use at least 512 bytes of disk space, the same as a file with 500 bytes or with 512 bytes. A file with 513 bytes will start using the next block, so it uses 1024 bytes of disk space.

So there is always some extent of difference between the sum of all the byte sizes of files and the sum of the blocks that are actually allocated (plus some meta data for i-nodes and sometimes journals, depending on the filesystem type).

Tools like the `df` command or QDirStat's _Filesystems_ window use the `statfs()` system call that returns data from the filesystem's internal _free blocks count_ and _total block count_; but those data are only available for the complete filesystem, not for subtrees.

The `du` command tries to be clever, taking the likely wasted partially used blocks at the end of files into account: So it adds 2 blocks (1024 bytes) for that 513 bytes file, not only the 513 bytes of its content.

QDirStat strictly uses the byte sizes, i.e. the 513 bytes in this example. This avoids having to deal with two different sizes, for the user as well internally as well as in the user interface.

If the filesystem mainly consists of tiny files, the difference between the sizes can up to a lot; Linux root filesystems are notorious for that because they use a _lot_ of tiny files:

![QDirStat-root-size-stat](https://user-images.githubusercontent.com/11538225/84182360-8f9f1e00-aa8a-11ea-8bc9-5675c4be59ef.png)

Notice the median of all those files is only 1.9 kB, just less than 4 blocks; so wasting on average half a block (256 bytes) per file makes a drastic difference: About 500 MB (9.4 GB - 8.9 GB) in this case. 

On the other hand, on my /work filesystem the

**Comment by shundhammer** [comment]
On the other hand, on my /work filesystem where I keep my photos, MP3s, videos, virtual machines and checked-out Git repos, the difference is hardly noticeable:

```
% sudo du -hsx /work
227G	/work

% df -h /work
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdb5       856G  227G  630G  27% /work

% sudo qdirstat /work
```

![QDirStat-work](https://user-images.githubusercontent.com/11538225/84182991-8bbfcb80-aa8b-11ea-9bd3-f9bce0ffcdc0.png)

![QDirStat-fs](https://user-images.githubusercontent.com/11538225/84183027-95e1ca00-aa8b-11ea-9047-d8ea245d502b.png)

The median is higher with 4.8 kB (i.e. about 9 blocks), but still in the same dimension (a one-digit kilobyte number):

![QDirStat-work-size-stat](https://user-images.githubusercontent.com/11538225/84183067-a003c880-aa8b-11ea-900e-d2687aa03f5d.png)

Yet, this makes the difference between the sizes pretty much vanish away: 226.2 GB vs. 226.7 GB.

**Comment by shundhammer** [comment]
I did experiment with using block-based sizes, but the result was most unsatisfactory: All of a sudden, there are two types of sizes to deal with, the byte sizes vs. the block-based sizes.

For a (short) while I thought I could get away with using the block-based size in the tree widget for directories and in the "Subtree" part of the details panel on the right side.

But then what does a user expect when he selects a number of small files in the tree or in the treemap? Should the sum of the byte sizes appear, or the sum of the block-based sizes? While probably nobody would notice the difference for normal-sized files, it becomes glaringly apparent for a bunch of tiny files: Selecting 10 files of 42 bytes each would no longer yield a sum of 420 bytes, but 5 kB (10 blocks allocated).

What is the user interested in? We can't know, so we'd have to present _both_ sizes somehow, cluttering the screen and confusing the user.

**Comment by shundhammer** [comment]
## Fragment Handling

But it gets worse: Some filesystem do support handling file _fragments_, storing them in a more efficient way; so the one extra byte of that 513 bytes file might or might not consume 2 blocks. Btrfs (and the now-obsolete Reiserfs / Reiserfs4) are known to do that.

## Symlinks

What about symlinks? According to the `stat()` system call, a symlink may consume 16 bytes for the path that it points to, but 0 blocks. Eh - what?

## Special Kernel Filesystems

What about special directories like `/dev`, `/proc`, `/sys` that are actually just kernel tables that are exported to the filesystem for convenient use? According to the `stat()` system call, their directories have a small byte size (but amazingly enough not a multiple of 512 bytes like normal directories), but they consume 0 blocks. Same with block and character devices. Eh - what?

## Bottom line

So that experiment with trying to be closer to the output of the `df` command by using block sizes only created confusion and little tangible benefit. I decided to abandon it.

There is no documented way to obtain information about fragment handling; the `du` command is basically just cheating by making assumptions.

It tends to be closer to the output of `df` and the `statfs()` system call, but only on certain filesystem types; `du` seems to guess correctly on ext4 etc. filesystems, but it fails horribly on NTFS and Btrfs.

So we'd gain a little more accuracy on ext4 at the price of having to deal with two different sizes in a lot of places, but the result would be just as off as it was before for other filesystem types.

**Comment by shundhammer** [comment]
_**Comments welcome!**_

**Comment by shundhammer** [comment]
## Allocated Sizes Reloaded

I gave it another try, and this appears to be much more promising.

There are subtle differences to the abandoned first attempt: Now it no longer tries to calculate the block-based allocated sizes on the fly; they are now calculated as each internal `FileInfo` object is created. Some filesystems can report the used blocks, some (e.g. the special kernel filesystems like `/dev`, `/proc`, `/sys`) can't. In that case, QDirStat now simply uses the byte size instead.

I experimented with adding the allocated sizes to the tree and gave that up after a very short while: The tree becomes a mess:

```
    1243 Bytes (4.0 KB allocated)
       17 Bytes (4.0 KB allocated)
```

Yikes.

So the tree keeps displaying only _one_ size for each item; it is the byte size for files and other non-directory items, and the allocated size for directories.

While this sounds pretty odd at first, it works amazingly well: It takes into account the general idea that a directory tree is a more abstract concept, dealing with more than just its individual parts. When I look at a subtree, I want to know how much overall disk space it consumes - the real thing, not just a theoretical value.

Breaking down the sizes into exact byte sizes (and their sum) and the allocated size(s) is now done in the "Details" panel on the right: It now has another row "Allocated" for each type of object, files as well as directories as well as the sum of currently selected items.

And if that "Allocated" field has the same content as its "Size" or "Total Size" counterpart, it displays nothing and is greyed out to minimize distraction.

This is particularly important for large files where the difference between byte size and allocated size is negligible; since QDirStat shows human-readable numbers, it would display "31.2 MB" vs. "31.2 MB". This would only confuse users, not help them.

![QDirStat-small-file](https://user-images.githubusercontent.com/11538225/84432909-c8c3c380-ac2d-11ea-865e-e5e8c38c75b2.png)

_Small file selected: Look at the "Details" panel. The allocated size does make a difference, so it makes sense to display it._

![QDirStat-large-file](https://user-images.githubusercontent.com/11538225/84432923-ccefe100-ac2d-11ea-933d-266a0686f39c.png)

_Large file selected: Half a block (256 Bytes) difference is beyond the granularity of the displayed size anyway, so no allocated size is shown._

**Comment by shundhammer** [comment]
Small files up to two clusters allocated size are now displayed like this:

![QDirStat-small-files-1](https://user-images.githubusercontent.com/11538225/84578692-ed53a300-adc7-11ea-83d9-169de04691dc.png)

Right now it's only like this if they use less than 75% of their allocated disk space.

The cluster size is determined from the first file of the directory tree with at least one block allocated and less than 512 bytes size.

As far as i tested, all major Linux filesystems have a 4k cluster size (ext4, Btrfs, XFS), and VFAT varies with the filesystem size; a 32 GB USB stick I tested has 16k clusters. No non-empty file can have less than one cluster allocated size (tested with all the above filesystems), so the alleged intelligent fragment handling for Btrfs appears to be a myth (or my Btrfs is simply too old).

**Comment by shundhammer** [comment]
Next iteration: Now displaying the allocated size more subdued, but for all small files that have 2 clusters or more (i.e. 8 KB) allocated. This is a lot less disruptive while scrolling through the list than sometimes seeing the "(4k)" or "(8k)" and sometimes not inbetween.

![QDirStat-small-file](https://user-images.githubusercontent.com/11538225/84600727-c060c800-ae7b-11ea-9a16-a5e3f14aef32.png)

It's amazing how insanely complicated Qt is for something as trivial as this: I had to create another ItemDelegate for that column and do low-level QPainter operations with QFontMetrics and manually calculating the new position for the substring. Surely in the year 2020 there must be an easier way to do this?!?

**Comment by theofficialgman** [comment]
I just want to comment here saying I was actually looking for this functionality myself!
as a user, it might benefit me to know where these tiny files that take up less than a block are so that I can compress/squash them and regain that disk space.
This whole thread gets a thumbs up from me
